src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
+                        outputMap[inputVar] = originalVar;
+                        var key = String.Format("{0}%{1}", originalVar, thisCBN.GUID);
+                        outputMap[key] = inputVar;
+                    }
+                }
+                else
+                {
+                    for (int i = 0; i < node.OutPorts.Count(); ++i)
+                    {
+                        var inputVar = node.GetAstIdentifierForOutputIndex(i).Value;
+                        outputMap[inputVar] = string.Empty;
+                    }
+                }
+            }
+
+            inputMap = inputVariableSet.ToDictionary(v => v, _ => string.Empty);
+        }
+
+        /// <summary>
+        /// Renumber variables used in astNode.
+        /// </summary>
+        /// <param name="astNode"></param>
+        /// <param name="numberingMap"></param>
+        /// <param name="variableMap"></param>
+        private static void VariableNumbering(
+            AssociativeNode astNode, 
+            NodeModel node,
+            Dictionary<string, NumberingState> numberingMap, 
+            Dictionary<string, string> renamingMap,
+            Dictionary<string, string> inputMap,
+            Dictionary<string, string> outputMap,
+            HashSet<string> mappedVariables)
+        {
+            Action<IdentifierNode> func = n =>
+            {
+                var ident = n.Value;
+
+                // This ident is from other node's output port, so it should be
+                // renamed later on VarialbeMapping()
+                if (renamingMap.ContainsKey(ident) || inputMap.ContainsKey(ident))
+                    return;
+
+                NumberingState ns; 
+                if (numberingMap.TryGetValue(ident, out ns))
+                {
+                    // ident already defined somewhere else. So we continue to 
+                    // bump its ID until numbered variable won't conflict with
+                    // all existing variables. 
+                    if (ns.IsNewSession)
+                    {
+                        ns.BumpID();
+                        while (mappedVariables.Contains(ns.NumberedVariable))
+                            ns.BumpID();
+
+                        ns.IsNewSession = false;
+                    }
+                    // ident already defined somewhere else, but we already
+                    // renumber this variable, so continue to use re-numbered
+                    // one.
+                }
+                else
+                {
+                    // It is a new variable. But we need to check if some other
+                    // variables are renamed to this one because of renumbering.
+                    // If there is confliction, continue to bump its ID.
+                    ns = new NumberingState(ident);
+                    numberingMap[ident] = ns;
+
+                    while (mappedVariables.Contains(ns.NumberedVariable))
+                        ns.BumpID();
+                }
```

Comment:
This is just me thinking out loud, you don't have to change if this doesn't make sense. I'm imagining that `NumberingState` class handles multiple variables internally instead of the current one variable. The `if-else` statement above can potentially be made this way:

```csharp
// NumberingState internally allocates entry for ident, and 
// adjust the identifier to accommodate for defined variables.
ns.AdjustForMappedVariable(ident, mappedVariables);
```

Because having the caller `VariableNumbering` method manages `numberingMap` might make it less readable. Consider if `VariableNumbering` can take just one `numberingState` object instead of a dictionary. But that can be quite a change.